src/algebra/homology/homology.lean

```diff
@@ -4,143 +4,179 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Scott Morrison, Markus Himmel
 -/
 import algebra.homology.chain_complex
-import category_theory.limits.shapes.images
-import category_theory.limits.shapes.kernels
+import algebra.homology.image_to_kernel_map
 
 /-!
-# Cohomology groups for cochain complexes
+# (Co)homology groups for complexes
 
-We setup that part of the theory of cohomology groups which works in
+We setup that part of the theory of homology groups which works in
 any category with kernels and images.
 
-We define the cohomology groups themselves, and show that they induce maps on kernels.
+We define the homology groups themselves, and show that they induce maps on kernels.
 
 Under the additional assumption that our category has equalizers and functorial images, we construct
-induced morphisms on images and functorial induced morphisms in cohomology.
+induced morphisms on images and functorial induced morphisms in homology.
 
+## Chains and cochains
+
+Throughout we work with complexes graded by an arbitrary `[add_comm_group β]`,
+with a differential with grading `b : β`.
+Thus we're simultaneously doing homology and cohomology groups
+(and in future, e.g., enabling computing homologies for successive pages of spectral sequences).
+
+At the end of the file we set up abbreviations `cohomology` and `cohomology_functor`,
+so that when you're working with a `C : cochain_complex V`, you can write `C.cohomology i`
+rather than the confusing `C.homology i`.
 -/
 
 universes v u
 
-namespace cochain_complex
-
 open category_theory
 open category_theory.limits
 
 variables {V : Type u} [category.{v} V] [has_zero_morphisms V]
 
-section
+variables {β : Type} [add_comm_group β] {b : β}
+
+namespace homological_complex
+
+section has_kernels
 
 variable [has_kernels V]
 
 /-- The map induced by a chain map between the kernels of the differentials. -/
-def kernel_map {C C' : cochain_complex V} (f : C ⟶ C') (i : ℤ) :
+def kernel_map {C C' : homological_complex V b} (f : C ⟶ C') (i : β) :
   kernel (C.d i) ⟶ kernel (C'.d i) :=
 kernel.lift _ (kernel.ι _ ≫ f.f i)
 begin
-  rw [category.assoc, ←comm_at f, ←category.assoc, kernel.condition, has_zero_morphisms.zero_comp],
+  rw [category.assoc, ←homological_complex.comm_at f, ←category.assoc, kernel.condition, has_zero_morphisms.zero_comp],
```

Comment:
Why do you need to change this? You are in the namespace and `comm_at` isn't protected, right?

Reply:
Good catch, reverted.